pkg/open: fix launching browser in some scenarios

fuhry opened the pull request:
Fleet Desktop may fail to launch a browser, or launch the wrong browser, in Ubuntu 24.04 LTS under some scenarios, including Wayland sessions and/or when the browser is installed as a snap or flatpak (by default, Ubuntu 24.04 LTS installs Firefox as a snap).

The previous approach of searching for an Xwayland process from which to extract environment variables is incomplete. When the browser is installed as a snap, `.desktop` files are installed by snapd into a location which is only discoverable through the `XDG_DATA_DIRS` environment variable.

To workaround this we need to change the approach in several ways:

- Look for `xdg-desktop-portal` processes, not just `Xwayland`. Both Firefox and Chrome support native operation under Wayland now, so it's not correct to rely upon the existence of an Xwayland process alone.
- Copy several more environment varibles: `XDG_CURRENT_DESKTOP`, `XDG_RUNTIME_DIR`, `XDG_DATA_DIRS`, and `PATH`, in addition to `XAUTHORITY`.

These variables allow fleet-desktop to honor the user's browser preference according to the currently running DE.

The list of variables to grab was found through trial-and-error and seems to be the minimum required for `xdg-open` to function.

Refactored and optimized the code to fetch environment variables from a running process to match only processes running as the current user and robustly handle multiple process results.

Fixes #19043, #27209.

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

<!-- Note that API documentation changes are now addressed by the product design team. -->

- [x] Changes file added for user-visible changes in `changes/`, `orbit/changes/` or `ee/fleetd-chrome/changes`.
  See [Changes files](https://github.com/fleetdm/fleet/blob/main/docs/Contributing/guides/committing-changes.md#changes-files) for more information.
- [ ] Manual QA for all new/changed functionality
- For Orbit and Fleet Desktop changes:
   - [ ] Manual QA must be performed in all relevant combinations of Linux desktop environments.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Bug Fixes**
  * Improved compatibility for launching browsers installed as Flatpak or Snap under Wayland sessions on Linux, resolving issues where Fleet Desktop failed to open these browsers.
  * Enhanced environment variable handling to better support browser launches in containerized and Wayland environments.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

fuhry commented:
# QA checklist

* [ ] Ubuntu, GNOME, Wayland session
* [ ] Ubuntu, GNOME, X11 session
* [ ] Ubuntu, KDE, Wayland session
* [ ] Ubuntu, KDE, X11 session
* [ ] Arch, Hyprland (Wayland only)

lucasmrod commented:
@fuhry Thanks for the contribution! This is on my list to review on the current sprint.

lucasmrod reviewed (commented):
Thanks for the contribution!

I've added some comments.

lucasmrod commented:
@fuhry: I've created https://github.com/fleetdm/fleet/issues/31087 to track the feature request as we currently don't support Snap/Flatpak installed browsers.

fuhry commented:
> @fuhry: I've created #31087 to track the feature request as we currently don't support Snap/Flatpak installed browsers.

Ack. Ubuntu 24.04 LTS installs Firefox as a snap by default, so this means a default (or pretty-close-to-default, i.e. the user picks "Unity on Wayland" at the GDM login screen) config can break.

I happened to find this while testing against flatpaks, but I believe the required changes are the same.

fuhry commented:
@lucasmrod I've addressed all comments, please re-review, thanks!

Simple reproducer:
* Install Ubuntu 24.04 LTS Desktop using default settings
* Install Orbit and reboot or logout
* At login screen, after selecting a user, click the cog icon in the lower-right corner and select "Ubuntu on Wayland"
* Click on "My Device" from the Fleet tray icon menu.

Using these steps, clicking "My Device" launches a word processor.

lucasmrod commented:
> I happened to find this while testing against flatpaks, but I believe the required changes are the same.

> Simple reproducer:
1. Install Ubuntu 24.04 LTS Desktop using default settings
2. Install Orbit and reboot or logout
3. At login screen, after selecting a user, click the cog icon in the lower-right corner and select "Ubuntu on Wayland"
4. Click on "My Device" from the Fleet tray icon menu.
Using these steps, clicking "My Device" launches a word processor.

Sorry, I'm a bit confused by these two comments.

I'm not able to reproduce with Firefox snap installed (default) on Ubuntu 24.04/25.04 (which uses Wayland by default).
Do any of the 1-4 steps above involve Flatpak installed Firefox?

Cool. I'll re-review the changes.

fuhry commented:
> I'm not able to reproduce with Firefox snap installed (default) on Ubuntu 24.04/25.04 (which uses Wayland by default).

What GPU is in your system? 24.04 seemed to default to X11 on my work-issued Dell Precision 5680 with an NVidia GPU, but I know that NVidia's Wayland support has come a long way in the past year. Maybe 25.04 is now defaulting to Wayland.

> Do any of the 1-4 steps above involve Flatpak installed Firefox?

No. I just reproduced this again on 24.04 with Firefox installed as a Snap.

fuhry commented on pkg/open/open_linux.go:
```diff
@@ -38,28 +56,99 @@ func browser(url string) error {
+				continue
+			}
+		}
+		if uids, err := p.Uids(); err == nil {
```
These come from `/proc/$PID/status`. There are 4 entries in the Uid and Gid row, representing real, effective, saved set and filesystem UUIDs. See: [proc_pid_status(5)](https://www.man7.org/linux/man-pages//man5/proc_pid_status.5.html#:~:text=Uid%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20Gid%20%20%20%20Real%2C%20effective%2C%20saved%20set%2C%20and%20filesystem%20UIDs%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20(GIDs).).

lucasmrod replied:
Ah TIL, thanks!

lucasmrod commented:
> > I'm not able to reproduce with Firefox snap installed (default) on Ubuntu 24.04/25.04 (which uses Wayland by default).
> 
> What GPU is in your system? 24.04 seemed to default to X11 on my work-issued Dell Precision 5680 with an NVidia GPU, but I know that NVidia's Wayland support has come a long way in the past year. Maybe 25.04 is now defaulting to Wayland.
> 

Lenovo Thinkpad L15 Gen 3 (AMD Radeon Graphics).
Ah, you most likely hit https://github.com/fleetdm/fleet/issues/27209 which we were never able to reproduce.
So, maybe you reproduced that issue and this PR is the fix. 
/cc @sgress454 

> > Do any of the 1-4 steps above involve Flatpak installed Firefox?
> 
> No. I just reproduced this again on 24.04 with Firefox installed as a Snap.

Gotcha.

lucasmrod commented on pkg/open/open_linux.go:
```diff
@@ -38,28 +56,114 @@ func browser(url string) error {
-		return "", fmt.Errorf("get executable path: %w", err)
+
+	var out []*gopsutil_process.Process
+	myUid := int32(os.Getuid())
```
You will need to dismiss the lint warning, e.g.:
https://github.com/fleetdm/fleet/blob/142c9e3d5b92ec971f18c5e30395d5ba56ba6a54/cmd/fleet/cron.go#L1442

Am guessing this is a-ok given this is a uid number.

fuhry replied:
Yes, that's correct. UIDs are never negative on any modern POSIX system that I'm aware of.

lucasmrod reviewed (commented):
Two small comments and it's g2g.

We will need to test this branch in Ubuntu 20.04 with X11 (our minimum supported version).

sharon-fdm commented:
@lucasmrod, @xpkoala, any update on this?

lukeheath commented:
@sharon-fdm @lucasmrod @xpkoala Converting to draft until we're ready to merge.  Please make sure to follow up.

sgress454 commented:
Hi @fuhry, thanks for your work on this. We did extensive testing and everything looked good with the exception of Kubuntu. On initial boot, selecting My Device from the desktop tray icon opens the expected page in Firefox. After that, it never works again, whether Firefox is open or not. Looking at the desktop error logs I see:
```
kmfclient: not found
test: Illegal number:
test: Illegal number:
test: Illegal number:
```
I can reproduce this in a terminal as root, using the same env and command line arguments as the desktop app uses (from looking at desktop logs). So something is not lining up with the env vars we're using. Specifically I tried:
```
sudo -u scott bash -c 'USER=scott SHLVL=0 XDG_CURRENT_DESKTOP=KDE XDG_RUNTIME_DIR=/run/user/1000 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus FLEET_DESKTOP_DEVICE_URL=http://redacted.com/device/79c57d62-f4a6-4636-a74b-06082a55f5bf FLEET_DESKTOP_INSECURE=1 SUDO_GID=0 XDG_SESSION_ID=c1 PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin MAIL=/var/mail/scott DISPLAY=:0 XAUTHORITY=/home/scott/.Xauthority XDG_SESSION_CLASS="session" TERM=unknown XDG_DATA_DIRS=/usr/share/plasma:/usr/local/share:/usr/share:/var/lib/snapd/desktop FLEET_DESKTOP_ALTERNATIVE_BROWSER_HOST= FLEET_DESKTOP_FLEET_URL=http://redacted.com FLEET_DESKTOP_FLEET_TLS_CLIENT_CERTIFICATE= FLEET_DESKTOP_TUF_UPDATE_ROOT=/opt/orbit xdg-open http://google.com'
```
and got those same errors. I tried changing `XDG_SESSION_CLASS` to `user` on a suggestion from chatgpt but neither that nor removing the var changed the outcome.  This was on a Kubuntu 22.04.04 VM on an x86-64 system.

I verified that the current production versions of orbit/orbit desktop do not have this issue.
